### DISCUSSION TITLE
Dynamic roles / several SaaS levels

### DISCUSSION THREAD

**Discussion by lumpidu** [zitadel/zitadel#6934] [Q&A]
I'd like to make roles configurable dynamically via callbacks. E.g. if a client/user upgrades it's SaaS plan, the returned roles or claims should also be updated in Zitadel.

Let's say a user starts in a **Free Plan**, so he/she can use the basic services. The SaaS service requests authentication/authorization from Zitadel if the user logs in/uses the API and does e.g. some rate limits based on the current user role/claims.

Then the user decides to upgrade to the **Pro Plan** at some later time and here e.g. the rate limits are increased. As soon as the plan upgrade is complete, he should immediate be able to benefit from it.

Besides the general implementation via Zitadel, I am also thinking about the "real-time" factor of this authorization: the user should be able to have immediate functionality of the new plan but authentication/authorization of every API request is too excessive for a high-volume API.
 
Which kind of technique/setup would you use to make this possible ?

**Comment by hifabienne** [comment]
Hi @lumpidu 
If you are using different roles on your project e.g freePlan, proPlan, etc and assigne them to the users I think that should possible.
As soon as the user upgrades you can call our APIs and change the role of the user. 
https://zitadel.com/docs/apis/resources/mgmt/management-service-add-user-grant

**Reply by lumpidu** [reply]
That sounds reasonable. Should my API then be registered as an API application or do I need a service account to make these Zitadel API calls ?

What about the API throughput ? I mean I can cache current access tokens of the user inside the API itself after a valid authentication/authorization from Zitadel and implement an invalidation call to my API for the customer after a plan upgrade/change. But I could also call Zitadel every time in sync to a call to my API is called.

In the latter case: how much time besides the typical ping time does this need ? I know that heavily depends on server specifications, but what ballpark are we talking about ? In the 10-30ms, 100-300ms or >1 seconds realm ?

**Reply by hifabienne** [reply]
You can use a service user to call the APIs.

It depends a little on your usecase, if you upgrade you can directly exchange the token. If you have cached tokens it depends on how long you are willing to allow users to have tokens that are not 100% uptodate. Especially when downgrading. Otherwise you can also react on certain errors and then create a new token.

It depends a little on how you implement it, if you are doing basig auth then it takes longer as the secret needs to be checked, in that case its longer than 1 second.  At the moment it always takes a little long to do so, but we are currently working on some performance improvements, the goal would be less than 100ms.

**Reply by lumpidu** [reply]
Ok, then I need probably to go with caching as my time constraints are much less than 1 second and even 100 ms is too much for my purposes. But that shouldn't be a big problem as the invalidation only needs to be done in a plan change which is relatively seldom and token invalidation can be done synchronously then.

I understand that I can use the service user and assign it the 'Org User Permission Editor'  rights so that it could manage the dynamic roles of the user. But I somehow don't really understand the API type of application: for what exactly would I need this kind of application type instead/additionally of/to the service user ?

**Reply by hifabienne** [reply]
reqeusts are always done by users either a human or a service user.
The api application is needed especially for oidc flows, for example if you have a frontend and a backend and the backend is called from the frontend. in oidc the audience for which the token is issued does have to be included in the tokens. in that case the audience of the backend is the client id of that app.

**Reply by lumpidu** [reply]
But where does the client secret come into play for the API application ? Is it used to verify user token credentials for the project it belongs to ?

**Reply by lumpidu** [reply]
Or maybe better asked: does it give the authorization to access the Zitadel management API to verify on a project-basis the authorization of the user roles ?

**Reply by hifabienne** [reply]
@livio-a can you help here?
